File: lsif/src/server/backend/database.test.ts
import * as sqliteModels from '../../shared/models/sqlite'
import { comparePosition, findRanges, mapRangesToInternalLocations, Database } from './database'
import * as fs from 'mz/fs'
import * as nodepath from 'path'
import * as constants from '../../shared/constants'
import { convertLsif } from '../../dump-processor/conversion/importer'
import { PathExistenceChecker } from '../../dump-processor/conversion/existence'
import rmfr from 'rmfr'
import * as uuid from 'uuid'
import { createStorageRoot } from '../../shared/test-util'

describe('Database', () => {
    let storageRoot!: string
    let database!: Database

    const makeDatabase = async (filename: string): Promise<Database> => {
        // Create a filesystem read stream for the given test file. This will cover
        // the cases where `yarn test` is run from the root or from the lsif directory.
        const root = (await fs.exists('lsif')) ? 'lsif' : ''
        const sourceFile = nodepath.join(root, 'test-data', filename)
        const databaseFile = nodepath.join(storageRoot, constants.TEMP_DIR, uuid.v4())

        await convertLsif({
            path: sourceFile,
            root: '',
            database: databaseFile,
            pathExistenceChecker: new PathExistenceChecker({
                repositoryId: 42,
                commit: 'ad3507cbeb18d1ed2b8a0f6354dea88a101197f3',
                root: '',
            }),
        })

        return new Database(1, databaseFile)
    }

    beforeAll(async () => {
        storageRoot = await createStorageRoot()
        database = await makeDatabase('lsif-go@ad3507cb.lsif.gz')
    })

    afterAll(async () => {
        if (storageRoot) {
            await rmfr(storageRoot)
        }
    })

    describe('exists', () => {
        it('should check document path', async () => {
            expect(await database.exists('cmd/lsif-go/main.go')).toEqual(true)
            expect(await database.exists('internal/index/indexer.go')).toEqual(true)
            expect(await database.exists('missing.go')).toEqual(false)
        })
    })

    describe('definitions', () => {
        it('should correlate definitions', async () => {
            // `\ts, err := indexer.Index()` -> `\t Index() (*Stats, error)`
            //                      ^^^^^           ^^^^^

            expect(await database.definitions('cmd/lsif-go/main.go', { line: 110, character: 22 })).toEqual([
                {
                    dumpId: 1,
                    path: 'internal/index/indexer.go',
                    range: { start: { line: 20, character: 1 }, end: { line: 20, character: 6 } },
                },
            ])
        })
    })

    describe('references', () => {
        it('should correlate references', async () => {
            // `func (w *Writer) EmitRange(start, end Pos) (string, error) {`
            //                   ^^^^^^^^^
            //
            // -> `\t\trangeID, err := i.w.EmitRange(lspRange(ipos, ident.Name, isQuotedPkgName))`
            //                             ^^^^^^^^^
            //
            // -> `\t\t\trangeID, err = i.w.EmitRange(lspRange(ipos, ident.Name, false))`
            //                              ^^^^^^^^^

            expect((await database.references('protocol/writer.go', { line: 85, character: 20 })).locations).toEqual([
                {
                    dumpId: 1,
                    path: 'protocol/writer.go',
                    range: { start: { line: 85, character: 17 }, end: { line: 85, character: 26 } },
                },
                {
                    dumpId: 1,
                    path: 'internal/index/indexer.go',
                    range: { start: { line: 529, character: 22 }, end: { line: 529, character: 31 } },
                },
                {
                    dumpId: 1,
                    path: 'internal/index/indexer.go',
                    range: { start: { line: 380, character: 22 }, end: { line: 380, character: 31 } },
                },
            ])
        })
    })

    describe('hover', () => {
        it('should correlate hover text', async () => {
            // `\tcontents, err := findContents(pkgs, p, f, obj)`
            //                     ^^^^^^^^^^^^

            const ticks = '```'
            const docstring = 'findContents returns contents used as hover info for given object.'
            const signature =
                'func findContents(pkgs []*Package, p *Package, f *File, obj Object) ([]MarkedString, error)'

            expect(await database.hover('internal/index/indexer.go', { line: 628, character: 20 })).toEqual({
                text: `${ticks}go\n${signature}\n${ticks}\n\n---\n\n${docstring}`,
                range: { start: { line: 628, character: 18 }, end: { line: 628, character: 30 } },
            })
        })
    })

    describe('getRangeByPosition', () => {
        it('should return correct range and document with monikers', async () => {
            // `func NewMetaData(id, root string, info ToolInfo) *MetaData {`
            //       ^^^^^^^^^^^

            const { document, ranges } = await database.getRangeByPosition('protocol/protocol.go', {
                line: 92,
                character: 10,
            })

            expect(ranges).toHaveLength(1)
            expect(ranges[0].startLine).toEqual(92)
            expect(ranges[0].startCharacter).toEqual(5)
            expect(ranges[0].endLine).toEqual(92)
            expect(ranges[0].endCharacter).toEqual(16)

            const monikers = Array.from(ranges[0].monikerIds).map(id => document?.monikers.get(id))
            expect(monikers).toHaveLength(1)
            expect(monikers[0]?.kind).toEqual('export')
            expect(monikers[0]?.scheme).toEqual('gomod')
            expect(monikers[0]?.identifier).toEqual('github.com/sourcegraph/lsif-go/protocol:NewMetaData')

            const packageInformation = document?.packageInformation.get(monikers[0]?.packageInformationId || 0)
            expect(packageInformation?.name).toEqual('github.com/sourcegraph/lsif-go')
            expect(packageInformation?.version).toEqual('v0.0.0-ad3507cbeb18')
        })
    })

    describe('monikerResults', () => {
        const edgeLocations = [
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 600, character: 1 }, end: { line: 600, character: 5 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 644, character: 1 }, end: { line: 644, character: 5 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 507, character: 1 }, end: { line: 507, character: 5 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 553, character: 1 }, end: { line: 553, character: 5 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 462, character: 1 }, end: { line: 462, character: 5 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 484, character: 1 }, end: { line: 484, character: 5 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 410, character: 5 }, end: { line: 410, character: 9 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 622, character: 1 }, end: { line: 622, character: 5 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 440, character: 1 }, end: { line: 440, character: 5 } },
            },
            {
                dumpId: 1,
                path: 'protocol/protocol.go',
                range: { start: { line: 530, character: 1 }, end: { line: 530, character: 5 } },
            },
        ]

        it('should query definitions table', async () => {
            const { locations, count } = await database.monikerResults(
                sqliteModels.DefinitionModel,
                {
                    scheme: 'gomod',
                    identifier: 'github.com/sourcegraph/lsif-go/protocol:Edge',
                },
                {}
            )

            expect(locations).toEqual(edgeLocations)
            expect(count).toEqual(10)
        })

        it('should respect pagination', async () => {
            const { locations, count } = await database.monikerResults(
                sqliteModels.DefinitionModel,
                {
                    scheme: 'gomod',
                    identifier: 'github.com/sourcegraph/lsif-go/protocol:Edge',
                },
                { skip: 3, take: 4 }
            )

            expect(locations).toEqual(edgeLocations.slice(3, 7))
            expect(count).toEqual(10)
        })

        it('should query references table', async () => {
            const { locations, count } = await database.monikerResults(
                sqliteModels.ReferenceModel,
                {
                    scheme: 'gomod',
                    identifier: 'github.com/slimsag/godocmd:ToMarkdown',
                },
                {}
            )

            expect(locations).toEqual([
                {
                    dumpId: 1,
                    path: 'internal/index/helper.go',
                    range: { start: { line: 78, character: 6 }, end: { line: 78, character: 16 } },
                },
            ])
            expect(count).toEqual(1)
        })
    })
})

describe('findRanges', () => {
    it('should return ranges containing position', () => {
        const range1 = {
            startLine: 0,
            startCharacter: 3,
            endLine: 0,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }
        const range2 = {
            startLine: 1,
            startCharacter: 3,
            endLine: 1,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }
        const range3 = {
            startLine: 2,
            startCharacter: 3,
            endLine: 2,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }
        const range4 = {
            startLine: 3,
            startCharacter: 3,
            endLine: 3,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }
        const range5 = {
            startLine: 4,
            startCharacter: 3,
            endLine: 4,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }

        expect(findRanges([range1, range2, range3, range4, range5], { line: 0, character: 4 })).toEqual([range1])
        expect(findRanges([range1, range2, range3, range4, range5], { line: 1, character: 4 })).toEqual([range2])
        expect(findRanges([range1, range2, range3, range4, range5], { line: 2, character: 4 })).toEqual([range3])
        expect(findRanges([range1, range2, range3, range4, range5], { line: 3, character: 4 })).toEqual([range4])
        expect(findRanges([range1, range2, range3, range4, range5], { line: 4, character: 4 })).toEqual([range5])
    })

    it('should order inner-most ranges first', () => {
        const range1 = {
            startLine: 0,
            startCharacter: 3,
            endLine: 4,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }
        const range2 = {
            startLine: 1,
            startCharacter: 3,
            endLine: 3,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }
        const range3 = {
            startLine: 2,
            startCharacter: 3,
            endLine: 2,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }
        const range4 = {
            startLine: 5,
            startCharacter: 3,
            endLine: 5,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }
        const range5 = {
            startLine: 6,
            startCharacter: 3,
            endLine: 6,
            endCharacter: 5,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }

        expect(findRanges([range1, range2, range3, range4, range5], { line: 2, character: 4 })).toEqual([
            range3,
            range2,
            range1,
        ])
    })
})

describe('comparePosition', () => {
    it('should return the relative order to a range', () => {
        const range = {
            startLine: 5,
            startCharacter: 11,
            endLine: 5,
            endCharacter: 13,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        }

        expect(comparePosition(range, { line: 5, character: 11 })).toEqual(0)
        expect(comparePosition(range, { line: 5, character: 12 })).toEqual(0)
        expect(comparePosition(range, { line: 5, character: 13 })).toEqual(0)
        expect(comparePosition(range, { line: 4, character: 12 })).toEqual(+1)
        expect(comparePosition(range, { line: 5, character: 10 })).toEqual(+1)
        expect(comparePosition(range, { line: 5, character: 14 })).toEqual(-1)
        expect(comparePosition(range, { line: 6, character: 12 })).toEqual(-1)
    })
})

describe('mapRangesToInternalLocations', () => {
    it('should map ranges to locations', () => {
        const ranges = new Map<sqliteModels.RangeId, sqliteModels.RangeData>()
        ranges.set(1, {
            startLine: 1,
            startCharacter: 1,
            endLine: 1,
            endCharacter: 2,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        })
        ranges.set(2, {
            startLine: 3,
            startCharacter: 1,
            endLine: 3,
            endCharacter: 2,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        })
        ranges.set(4, {
            startLine: 2,
            startCharacter: 1,
            endLine: 2,
            endCharacter: 2,
            monikerIds: new Set<sqliteModels.MonikerId>(),
        })

        const path = 'src/position.ts'
        const locations = mapRangesToInternalLocations(42, ranges, path, new Set([1, 2, 4]))
        expect(locations).toContainEqual({
            dumpId: 42,
            path,
            range: { start: { line: 1, character: 1 }, end: { line: 1, character: 2 } },
        })
        expect(locations).toContainEqual({
            dumpId: 42,
            path,
            range: { start: { line: 3, character: 1 }, end: { line: 3, character: 2 } },
        })
        expect(locations).toContainEqual({
            dumpId: 42,
            path,
            range: { start: { line: 2, character: 1 }, end: { line: 2, character: 2 } },
        })
        expect(locations).toHaveLength(3)
    })
})

File: cmd/precise-code-intel-bundle-manager/internal/database/database_test.go
package database

import (
	"context"
	"testing"

	"github.com/google/go-cmp/cmp"
	sqlitereader "github.com/sourcegraph/sourcegraph/internal/codeintel/bundles/persistence/sqlite"
	"github.com/sourcegraph/sourcegraph/internal/codeintel/bundles/types"
	"github.com/sourcegraph/sourcegraph/internal/observation"
	"github.com/sourcegraph/sourcegraph/internal/sqliteutil"
)

func init() {
	sqliteutil.SetLocalLibpath()
	sqliteutil.MustRegisterSqlite3WithPcre()
}

func TestDatabaseExists(t *testing.T) {
	testCases := []struct {
		path     string
		expected bool
	}{
		{"cmd/lsif-go/main.go", true},
		{"internal/index/indexer.go", true},
		{"missing.go", false},
	}

	db := openTestDatabase(t)
	for _, testCase := range testCases {
		if exists, err := db.Exists(context.Background(), testCase.path); err != nil {
			t.Fatalf("unexpected error %s", err)
		} else if exists != testCase.expected {
			t.Errorf("unexpected exists result for %s. want=%v have=%v", testCase.path, testCase.expected, exists)
		}
	}
}

func TestDatabaseDefinitions(t *testing.T) {
	// `\ts, err := indexer.Index()` -> `\t Index() (*Stats, error)`
	//                      ^^^^^           ^^^^^

	db := openTestDatabase(t)
	if actual, err := db.Definitions(context.Background(), "cmd/lsif-go/main.go", 110, 22); err != nil {
		t.Fatalf("unexpected error %s", err)
	} else {
		expected := []Location{
			{
				Path:  "internal/index/indexer.go",
				Range: newRange(20, 1, 20, 6),
			},
		}

		if diff := cmp.Diff(expected, actual); diff != "" {
			t.Errorf("unexpected definitions locations (-want +got):\n%s", diff)
		}
	}
}

func TestDatabaseReferences(t *testing.T) {
	// `func (w *Writer) EmitRange(start, end Pos) (string, error) {`
	//                   ^^^^^^^^^
	//
	// -> `\t\trangeID, err := i.w.EmitRange(lspRange(ipos, ident.Name, isQuotedPkgName))`
	//                             ^^^^^^^^^
	//
	// -> `\t\t\trangeID, err = i.w.EmitRange(lspRange(ipos, ident.Name, false))`
	//                              ^^^^^^^^^

	db := openTestDatabase(t)
	if actual, err := db.References(context.Background(), "protocol/writer.go", 85, 20); err != nil {
		t.Fatalf("unexpected error %s", err)
	} else {
		expected := []Location{
			{
				Path:  "internal/index/indexer.go",
				Range: newRange(529, 22, 529, 31),
			}, {
				Path:  "internal/index/indexer.go",
				Range: newRange(380, 22, 380, 31),
			},
			{
				Path:  "protocol/writer.go",
				Range: newRange(85, 17, 85, 26),
			},
		}

		if diff := cmp.Diff(expected, actual); diff != "" {
			t.Errorf("unexpected reference locations (-want +got):\n%s", diff)
		}
	}
}

func TestDatabaseHover(t *testing.T) {
	// `\tcontents, err := findContents(pkgs, p, f, obj)`
	//                     ^^^^^^^^^^^^

	db := openTestDatabase(t)
	if actualText, actualRange, exists, err := db.Hover(context.Background(), "internal/index/indexer.go", 628, 20); err != nil {
		t.Fatalf("unexpected error %s", err)
	} else if !exists {
		t.Errorf("no hover found")
	} else {
		docstring := "findContents returns contents used as hover info for given object."
		signature := "func findContents(pkgs []*Package, p *Package, f *File, obj Object) ([]MarkedString, error)"
		expectedText := "```go\n" + signature + "\n```\n\n---\n\n" + docstring
		expectedRange := newRange(628, 18, 628, 30)

		if actualText != expectedText {
			t.Errorf("unexpected hover text. want=%s have=%s", expectedText, actualText)
		}

		if diff := cmp.Diff(expectedRange, actualRange); diff != "" {
			t.Errorf("unexpected hover range (-want +got):\n%s", diff)
		}
	}
}

func TestDatabaseMonikersByPosition(t *testing.T) {
	// `func NewMetaData(id, root string, info ToolInfo) *MetaData {`
	//       ^^^^^^^^^^^

	db := openTestDatabase(t)
	if actual, err := db.MonikersByPosition(context.Background(), "protocol/protocol.go", 92, 10); err != nil {
		t.Fatalf("unexpected error %s", err)
	} else {
		expected := [][]types.MonikerData{
			{
				{
					Kind:                 "export",
					Scheme:               "gomod",
					Identifier:           "github.com/sourcegraph/lsif-go/protocol:NewMetaData",
					PackageInformationID: types.ID("213"),
				},
			},
		}

		if diff := cmp.Diff(expected, actual); diff != "" {
			t.Errorf("unexpected moniker result (-want +got):\n%s", diff)
		}
	}
}

func TestDatabaseMonikerResults(t *testing.T) {
	edgeLocations := []Location{
		{
			Path:  "protocol/protocol.go",
			Range: newRange(600, 1, 600, 5),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(644, 1, 644, 5),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(507, 1, 507, 5),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(553, 1, 553, 5),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(462, 1, 462, 5),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(484, 1, 484, 5),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(410, 5, 410, 9),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(622, 1, 622, 5),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(440, 1, 440, 5),
		},
		{
			Path:  "protocol/protocol.go",
			Range: newRange(530, 1, 530, 5),
		},
	}

	markdownLocations := []Location{
		{
			Path:  "internal/index/helper.go",
			Range: newRange(78, 6, 78, 16),
		},
	}

	testCases := []struct {
		tableName          string
		scheme             string
		identifier         string
		skip               int
		take               int
		expectedLocations  []Location
		expectedTotalCount int
	}{
		{"definitions", "gomod", "github.com/sourcegraph/lsif-go/protocol:Edge", 0, 100, edgeLocations, 10},
		{"definitions", "gomod", "github.com/sourcegraph/lsif-go/protocol:Edge", 3, 4, edgeLocations[3:7], 10},
		{"references", "gomod", "github.com/slimsag/godocmd:ToMarkdown", 0, 100, markdownLocations, 1},
	}

	db := openTestDatabase(t)
	for i, testCase := range testCases {
		if actual, totalCount, err := db.MonikerResults(context.Background(), testCase.tableName, testCase.scheme, testCase.identifier, testCase.skip, testCase.take); err != nil {
			t.Fatalf("unexpected error for test case #%d: %s", i, err)
		} else {
			if totalCount != testCase.expectedTotalCount {
				t.Errorf("unexpected moniker result total count for test case #%d. want=%d have=%d", i, testCase.expectedTotalCount, totalCount)
			}

			if diff := cmp.Diff(testCase.expectedLocations, actual); diff != "" {
				t.Errorf("unexpected moniker result locations for test case #%d (-want +got):\n%s", i, diff)
			}
		}
	}
}

func TestDatabasePackageInformation(t *testing.T) {
	db := openTestDatabase(t)
	if actual, exists, err := db.PackageInformation(context.Background(), "protocol/protocol.go", types.ID("213")); err != nil {
		t.Fatalf("unexpected error %s", err)
	} else if !exists {
		t.Errorf("no package information")
	} else {
		expected := types.PackageInformationData{
			Name:    "github.com/sourcegraph/lsif-go",
			Version: "v0.0.0-ad3507cbeb18",
		}

		if diff := cmp.Diff(expected, actual); diff != "" {
			t.Errorf("unexpected package information (-want +got):\n%s", diff)
		}
	}
}

func openTestDatabase(t *testing.T) Database {
	filename := "../../../../internal/codeintel/bundles/persistence/sqlite/testdata/lsif-go@ad3507cb.lsif.db"

	// TODO(efritz) - rewrite test not to require actual reader
	reader, err := sqlitereader.NewReader(context.Background(), filename)
	if err != nil {
		t.Fatalf("unexpected error creating reader: %s", err)
	}

	documentCache, _, err := NewDocumentCache(1)
	if err != nil {
		t.Fatalf("unexpected error creating cache: %s", err)
	}

	resultChunkCache, _, err := NewResultChunkCache(1)
	if err != nil {
		t.Fatalf("unexpected error creating cache: %s", err)
	}

	db, err := OpenDatabase(context.Background(), filename, reader, documentCache, resultChunkCache)
	if err != nil {
		t.Fatalf("unexpected error opening database: %s", err)
	}
	t.Cleanup(func() { _ = db.Close })

	// Wrap in observed, as that's how it's used in production
	return NewObserved(db, filename, &observation.TestContext)
}
